app/src/js/server/user_manager.ts

```diff
@@ -19,106 +20,159 @@ export class UserManager {
 
     // Update user metadata
     const metaKey = getMetaKey()
-    let socketToProject: { [key: string]: string } = {}
-    if (await this.storage.hasKey(metaKey)) {
-      const metaDataJSON = await this.storage.load(metaKey)
-      socketToProject = safeParseJSON(metaDataJSON)
-    }
+    const socketToProject = await this.getSocketToProjectMap(metaKey)
     socketToProject[socketId] = projectName
-    await this.storage.save(metaKey, JSON.stringify(socketToProject))
+    await this.saveMeta(metaKey, socketToProject)
   }
 
   /**
    * Deletes the user data of the socket that disconnected
    */
   public async deregisterUser (socketId: string) {
-    // First access the projectName via metadata
+    // Access the projectName via metadata
     const metaKey = getMetaKey()
-    if (!(await this.storage.hasKey(metaKey))) {
-      return
-    }
-    const metaDataJSON = await this.storage.load(metaKey)
-    const socketToProject = safeParseJSON(metaDataJSON)
+    const socketToProject = await this.getSocketToProjectMap(metaKey)
     if (!(socketId in socketToProject)) {
+      // socket has no associated project
       return
     }
     const projectName = socketToProject[socketId]
 
+    // Remove the socket info from the metadata
+    delete socketToProject[socketId]
+    await this.saveMeta(metaKey, socketToProject)
+
     // Next remove the user info for that project
     const key = getUserKey(projectName)
-    if (!(await this.storage.hasKey(key))) {
-      return
-    }
-    const userDataJSON = await this.storage.load(key)
-    const userData = safeParseJSON(userDataJSON)
+    const userData = await this.loadUserData(key)
     const socketToUser = userData.socketToUser
     const userToSockets = userData.userToSockets
 
     if (!socketToUser || !(socketId in socketToUser)) {
+      // socket has no associated user
       return
     }
+    // Remove map from socket to user
     const userId = socketToUser[socketId]
+    delete socketToUser[socketId]
+
+    // Remove socket from its user's list
     const socketInd = userToSockets[userId].indexOf(socketId)
     if (socketInd > -1) {
-      // remove the socket from the user
       userToSockets[userId].splice(socketInd, 1)
     }
+
+    // Remove the user if it has no sockets left
     if (userToSockets[userId].length === 0) {
       delete userToSockets[userId]
     }
+    await this.saveUser(key, socketToUser, userToSockets)
+  }
 
-    delete socketToUser[socketId]
+  /**
+   * Counts the number of currently connected users
+   */
+  public async countUsers (projectName: string): Promise<number> {
+    const key = getUserKey(projectName)
+    const userData = await this.loadUserData(key)
+    const userToSockets = userData.userToSockets
+    if (!userToSockets) {
+      return 0
+    }
+    return Object.keys(userToSockets).length
+  }
 
-    const newUserData = {
+  /**
+   * Remove all active users, so all counts should be 0
+   */
+  public async clearUsers (): Promise<void> {
```

Comment:
Can you document the folder structure where the user info is stored?

Reply:
I added comments documenting this in the ProjectStore methods for saving/loading user data and metadata